djangocms_link/forms.py

```diff
@@ -37,11 +39,57 @@ def __init__(self, *args, **kwargs):
 
     def clean(self):
         cleaned_data = super(LinkForm, self).clean()
-        external_link = cleaned_data.get('external_link')
-        internal_link = cleaned_data.get('internal_link')
-        mailto = cleaned_data.get('mailto')
-        phone = cleaned_data.get('phone')
-        anchor = cleaned_data.get('anchor')
-        if not any([external_link, internal_link, mailto, phone, anchor]):
-            raise ValidationError(_('At least one link is required.'))
+        field_names = (
+            'external_link',
+            'internal_link',
+            'mailto',
+            'phone',
+        )
+        anchor_field_name = 'anchor'
+        field_names_allowed_with_anchor = (
+            'external_link',
+            'internal_link',
+        )
+
+        anchor_field_verbose_name = force_text(
+            getattr(self.fields.get(anchor_field_name), 'label'))
+        anchor_field_value = cleaned_data.get(anchor_field_name)
+
+        link_fields = {
+            key: cleaned_data.get(key)
+            for key in field_names
+        }
+        link_field_verbose_names = {
+            key: force_text(getattr(self.fields.get(key), 'label'))
+            for key in link_fields.keys()
+        }
+        provided_link_fields = {
+            key: value
+            for key, value in link_fields.items()
+            if value
+        }
+
+        if len(provided_link_fields) > 1:
+            # Too many fields have a value.
+            error_msg = '{} {}'.format(
+                _('Only one of these fields is allowed:'),
+                ', '.join(link_field_verbose_names.values()),
+            )
+            errors = {}
+            for field, value in link_fields.items():
+                if value:
+                    errors[field] = error_msg
+            raise ValidationError(errors)
+
+        if anchor_field_value:
+            for field_name in provided_link_fields.keys():
+                if field_name not in field_names_allowed_with_anchor:
+                    error_msg = _('%(anchor_field_verbose_name)s is not allowed together with %(field_name)s') % {
```

Comment:
I would define the message above without the formatting to avoid such long line.